xformers/components/attention/core.py

```diff
@@ -199,25 +201,26 @@ def scaled_query_key_softmax(
     q = q / math.sqrt(k.size(-1))
 
     # Matmul with mask, if boolean
-    is_bool_mask = att_mask is not None and att_mask.dtype == torch.bool
-    att = _matmul_with_mask(q, k.transpose(-2, -1), att_mask if is_bool_mask else None)
+    if att_mask is not None and isinstance(att_mask, AttentionMask):
+        # Additive mask
+        mask: Optional[Union[SparseCS, torch.Tensor]] = att_mask.values
+    else:
+        mask = att_mask
 
-    # Could also be that the mask was additive
-    if att_mask is not None and att_mask.dtype != torch.bool:
-        att = att + att_mask
+    att = _matmul_with_mask(q, k.transpose(-2, -1), mask)
 
     # Softmax to get the attention probabilities
-    att = _softmax(att, causal=causal)
+    is_causal = att_mask.is_causal if isinstance(att_mask, AttentionMask) else False
```

Comment:
this fixes https://github.com/facebookresearch/xformers/issues/91

Reply:
Thanks!

One downside of this approach is that it leaks internal implementation details of `AttentionMask` inside a core function.

My thoughts were to modify instead `_softmax` so that the causal path is just a different function call from a `CausalAttentionMask` Tensor type.

Reply:
in that case you pass the mask to softmax, it moves the branching there, why not. Typically softmax do not take masks though, so I think that the interface becomes a little confusing (we'll pass a mask in only to say that it's causal or not)

Reply:
@fmassa is that ok if you propose a change (object oriented if I got it right, the masks handle the dispatch problem) in another PR ?

Reply:
Sure, I'll send a PR with my proposed changes later this week